src/lib.rs

```diff
@@ -3789,6 +3801,431 @@ pub struct SemanticHighlightingParams {
     pub lines: Vec<SemanticHighlightingInformation>,
 }
 
+/**
+ * A set of predefined token types. This set is not fixed
+ * an clients can specify additional token types via the
+ * corresponding client capabilities.
+ *
+ * @since 3.16.0 - Proposed state
+ */
+#[derive(Debug, Eq, PartialEq, Clone, Copy, Deserialize, Serialize)]
+#[cfg(feature = "proposed")]
+pub enum SemanticTokenTypes {
```

Comment:
I might be missing something, but what's the purpose of this being `Serialize/Deserialize` at all? This is not used in any message, it's just a **sub**set of well-known types. 

Perhaps we need something like this instead:

```rust
pub struct SemanticTokenType(pub String);

impl SemanticTokenType {
    fn comment() -> Self { Self("comment".to_owned()) } // perhaps generate with a macro
    ....
}

pub struct SemanticTokenModifier(pub String);

pub struct SemanticTokensLegend {
    pub token_types: Vec<SemanticTokenType>,
    pub token_modifiers: Vec<SemanticTokenModifier>,
}
```

Reply:
`#[serde(other)]` with support for one element variants would be ideal for this if it were implemented https://github.com/serde-rs/serde/issues/1701

Reply:
I am actually not sure... It's not really an enum, it's an unrestricted set of strings, some of which are well known. Ideally, what we need here is

```rust
impl SemanticTokenType {
    const COMMENT: &SemanticTokenType = &SemanticTokenType(String::from("comment"));
}
```

then clients can "extend" this

```rust
mod rust_token_types {
    const REFERENCE: &SemanticTokenType = &SemanticTokenType(String::from("reference"));
}
```

That is, TypeScript uses an `enum` just to namespace a bunch of string constants, not to introduce a new type.

Reply:
Anyway, I have no strong opinion on how this should look like, I am ok with any soln :)

Reply:
```rust
struct SemanticTokenType(Cow<'static, str>);
impl SemanticTokenType {
    const COMMENT: SemanticTokenType = SemanticTokenType(Cow::Borrowed("comment"));
}
```

is probably ideal in that case.

Reply:
Ohh, that's neat, I suggest the following API then:

```rust
// `Cow` is private, because it is kind of awkward type to expose
#[derive(Serialzie, Deserialize, Eq, Ord, Hash, Clone)]
pub struct SemanticTokenType(Cow<'static, str>); 

impl SemanticTokenType {
    const COMMENT: SemanticTokenType = SemanticTokenType::new("comment");

    // For servers, I expect al tags to be 'static in practice. 
    pub const fn new(tag: &'static str) -> SemanticTokenType { ... }

    // To inspect the contents
    pub fn as_str(&self) -> &'str { ... }
}
// So that you can construct from Strings
impl From<String> for SemanticTokenType;
// For symmetry
impl From<&'static str> for SemanticTokenType;
```

Another alternative is to be lazy and just use pub: `pub struct SemanticTokenType(pub Cow<'static, str>); `.

Reply:
The one downside would be that the constants can't be pattern matched on. I think that is (part of) why the `http` crate defines `Method` like https://docs.rs/http/0.2.0/src/http/method.rs.html#113-234

Reply:
Right. I am not sure that people would want to pattern match often in this case though: unlike HTTP methods, all tags should probably be handled in the same way, by applying some styling based on the config.

Reply:
I think I will try the `Cow` wrapper first and see how that shakes out?

Reply:
It doesn't look like `Cow` can be `const` on stable right now.

Reply:
I think it does? https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bf188c3cd3fc14bd9b1ec4df88e45e6c

Reply:
This is what I get for switching between C# and rust.